### ISSUE TITLE
Detect Mistaken Use Of Middleware Factories

### ISSUE THREAD

**Issue by hipitihop** [Day8/re-frame#65]
A handler with faulty middleware specification silently fails to fire without error or console warning. The following unit test shows a handler using the `undoable` middleware factory incorrectly.

```
(ns test.base.faulty-middleware
  (:require
    [cemerick.cljs.test :refer-macros [is deftest use-fixtures testing]]
    [re-frame.core      :refer [trim-v undoable]]
    [re-frame.db        :refer [app-db]]))


(defn setup []
  (reset! app-db {}))

(defn teardown []
  (reset! app-db {}))

(defn each-fixture [f]
  (setup)
  (f)
  (teardown))

(use-fixtures :each each-fixture)

; register dummy handler with faulty middleware.
; undoable is a factory s.b. (undoable "description")
(re-frame.core/register-handler
  :with-faulty-middleware
  [trim-v undoable]
  (fn [db [arg]] (assoc db :event-value arg)))

;; ---------- TESTS ------------------------------------------------------------

(deftest test-faulty-middleware
  (testing "dispatch to handler with faulty middleware should not silently fail"
    (re-frame.core/dispatch-sync [:with-faulty-middleware "event value"])
    (is (= "event value" (:event-value @app-db)) "Handler failed to fire properly and set :event-value.")))
```

**Comment by mike-thompson-day8** [comment]
**Problem Definition**
If you:
   - mistakenly supplying `undoable` as middleware ...
   - ... instead of correctly supplying this `(undoable "some explanation")`

`undoable` is a middleware factory.  You must call it (with a parameter) to get middleware. 

So this ticket is asking for the mistake to be reported because its a mistake which currently passes silently, and it is a bit baffling to debug.

**Solutions:** 
Unfortunately `undoable` is a function (middleware factory), just like a true middleware function would be.  So detection of this mistake is a bit hard. 

Doubly so because you can do this `(undoable some-function)` .  So we can't even check at middleware composition time for a non fn parameter.

Will dwell on it.

**Comment by mike-thompson-day8** [comment]
Okay, I've figured out a way of doing this via metadata.  But I'm stuck on this question:
https://groups.google.com/forum/#!topic/clojurescript/KuGrBMqzO1o

**Comment by ul** [comment]
not sure which answer will arrive faster to you, so duplicate here and in cljs group

defn attaches metadata to var:

```
(defn ^{:a 1} f [] true)
#<function cljs$user$f(){
return true;
}>
=> nil

(meta #'f)
=> {:ns cljs.user, :name f, :file "<cljs repl>", :end-column 16, :source "^{:a 1} f", :column 1, :line 1, :end-line 1, :arglists ([]), :doc nil, :test nil, :a 1}
```

note presence of `:a` key in the end

**Comment by mike-thompson-day8** [comment]
Fixed in 78dacf5ca1551b25b53cc527fbe4b26d3de1ceaf

Will be a part of v0.4.1
